2017/ru/0xa1-injection.md

```diff
@@ -0,0 +1,77 @@
+* вредоносные данные используются в поисковых параметрах объектно-реляционного отображения для извлечения дополнительной, критичной информации;
+* вредоносные данные используются или добавляются т.о., что SQL-код или команды содержат структурные и вредоносные данные в динамических запросах, командах или хранимых процедурах.
+
+Наиболее распространенными являются SQL-, NoSQL-, ORM-, LDAP-, EL- или OGNL-внедрения, а также внедрения команд ОС. То же самое касается всех интерпретаторов. Анализ исходного кода является лучшим способом обнаружения внедрений, за которым следует полное автоматизированное тестирование всех вводимых параметров, заголовков, URL, куки, JSON-, SOAP- и XML-данных. Организации также могут включать в процесс непрерывной интеграции и развертывания ПО (CI/CD) статическое ([SAST](https://www.owasp.org/index.php/Source_Code_Analysis_Tools)) и динамическое ([DAST](https://www.owasp.org/index.php/Category:Vulnerability_Scanning_Tools)) тестирование кода и приложений для обнаружения новых уязвимостей перед внедрением приложений в производство.
+<details>
+    <summary>
+        <b><i><font size="2" color="aqua">ORM | LDAP | EL | OGNL</font></i></b>
+    </summary>
+    <font size="2" color="aquamarine">
+        <b>ORM</b> → Объектно-реляционное отображение <br>
+        <b>LDAP</b> → Облегченный протокол доступа к каталогам<br>
+        <b>EL</b> → Язык выражений<br>
+        <b>OGNL</b> → Объектно-графовый язык навигации
+    </font>
+</details>
+
+## Как предотвратить
+
+Для предотвращения внедрений необходимо изолировать данные от команд и запросов.
+
+* Используйте безопасный API, исключающий применение интерпретатора или предоставляющий параметризованный интерфейс, либо используйте инструменты объектно-реляционного отображения (ORM).
+__Примечание__: даже параметризованные хранимые процедуры могут привести к SQL-внедрениям, если PL/SQL или T-SQL позволяют присоединять запросы и данные или выполнять вредоносный код с помощью EXECUTE IMMEDIATE или exec().
+* Реализуйте на сервере белые списки для проверки входных данных. Это, конечно, не обеспечит полную защиту, поскольку многие приложения используют спецсимволы, например, в текстовых областях или API для мобильных приложений.
+* Для остальных динамических запросов реализуйте экранирование спецсимволов, используя соответствующий интерпретатору синтаксис.
+__Примечание__: элементы SQL-структуры, такие как названия таблиц или столбцов, нельзя экранировать, поэтому предоставляемые пользователями названия представляют опасность. Это обычная проблема программ для составления отчетов.
+* Используйте в запросах LIMIT или другие элементы управления SQL для предотвращения утечек данных.
+
+## Примеры сценариев атак
+
+**Сценарий №1**: Приложение использует недоверенные данные при создании следующего уязвимого SQL-вызова:
+
+`String query = "SELECT * FROM accounts WHERE custID='" + request.getParameter("id") + "'";`
```

Comment:
It makes sense to use more concrete highlighting here. I think it's java:

````markdown
```java
String query = "SELECT * FROM accounts WHERE custID='" + request.getParameter("id") + "'";
```
````

Reply:
It's better to refer to the [original document](https://github.com/OWASP/Top10/blob/master/2017/OWASP%20Top%2010-2017%20(en).pdf).